k2/csrc/ragged_ops_inl.h

```diff
@@ -434,6 +434,158 @@ Ragged<T> Index(Ragged<T> &src, Ragged<int32_t> &indexes, bool remove_axis) {
+template <typename T>
+struct PairOutputIterator {  // outputs just the index of the pair.
+  // most of the following typedefs won't really be used; they are added because
+  // cub uses iterator_traits, which requires them.
+  typedef Pair<T> value_type;
+  typedef size_t difference_type;
+  typedef void pointer;
+  typedef PairOutputIteratorDeref<T> reference;
+  typedef std::output_iterator_tag iterator_category;
+
+  PairOutputIterator(int32_t *i) : i_(i), offset_(0) { }
+  __device__ PairOutputIterator(int32_t *i, size_t offset) :
+      i_(i), offset_(offset) { }
+  __device__ PairOutputIterator(const PairOutputIterator &other) :
+      i_(other.i_), offset_(other.offset_) { }
+  __device__ PairOutputIteratorDeref<T> operator[](int32_t idx) const {
+    return PairOutputIteratorDeref<T>(i_ + idx);
+  }
+  __device__ PairOutputIterator operator+(size_t offset) {
+    return PairOutputIterator { i_, offset_ + offset };
+  }
+
+  int32_t *i_;
+  size_t offset_;
+};
+
+template <typename T>
+struct PairMaxOp {
+  __device__ Pair<T> operator() (const Pair<T> &a,
+                                 const Pair<T> &b) const {
+    // NOTE: could specialize this via a union, if T == int32_t, might be
+    // marginally faster.
+    if (a.t > b.t || (a.t == b.t && a.idx > b.idx))
+      return a;
+    else return b;
+  }
+  PairMaxOp() { }
+  __host__ __device__ PairMaxOp(PairMaxOp &) { }
+};
+
+
+}
+
+
+template <typename T>
+void ArgMaxPerSublist(Ragged<T> &src, T initial_value, Array1<int32_t> *dst) {
+  NVTX_RANGE(K2_FUNC);
+  K2_CHECK_GE(src.NumAxes(), 2);
+  K2_CHECK(IsCompatible(src.shape, *dst));
+
+  int32_t last_axis = src.NumAxes() - 1;
+  const Array1<int32_t> &row_splits_array = src.RowSplits(last_axis);
+  int32_t num_rows = row_splits_array.Dim() - 1;
+  K2_CHECK_EQ(num_rows, dst->Dim());
+
+  ContextPtr &c = src.Context();
+  const int32_t *row_splits = row_splits_array.Data();
+  const T *values_data = src.values.Data();
+  int32_t *output_data = dst->Data();
+
+  if (c->GetDeviceType() == kCpu) {
+    int32_t j = row_splits[0];
+    for (int32_t i = 0; i < num_rows; ++i) {
+      T val = initial_value;
+      int32_t idx = -1;
+
+      int32_t row_end = row_splits[i + 1];
+      for (; j < row_end; ++j) {
+        T elem = values_data[j];
+        if (elem >= val) {
+          val = elem;
+          idx = j;
+        }
+      }
+      output_data[i] = idx;
+    }
+  } else {
+    K2_CHECK_EQ(c->GetDeviceType(), kCuda);
+    using namespace argmax_internal;
+
+    PairInputIterator<T> input_iter(values_data);
+    PairOutputIterator<T> output_iter(output_data);
+    PairMaxOp<T> op;
+    Pair<T> initial_pair { initial_value, -1 };
+
+    // This code is based on the example here:
+    // https://nvlabs.github.io/cub/structcub_1_1_device_segmented_reduce.html
+    std::size_t temp_storage_bytes = 0;
+
+    // the first time is to determine temporary device storage requirements
+    K2_CUDA_SAFE_CALL(cub::DeviceSegmentedReduce::Reduce(
```

Comment:
Just a question, wondering if we need the max values as well, if so, maybe we can call `cub::DeviceSegmentedReduce::ArgMax` https://nvlabs.github.io/cub/structcub_1_1_device_segmented_reduce.html#a2c61f69772a76c383ed8d34c6e5a2194 directly

Reply:
We'd need to define an appropriate pair type, which would be a slight hassle.  In general you can easily look up the max value. (I considered that approach.)

BTW the current code, I realized, is redundant: the output iterator stores an offset, which is not needed (can just move the pointer).  Perhaps you could fix that?

Reply:
OK, will do